cirq/contrib/quimb/density_matrix.py

```diff
@@ -0,0 +1,195 @@
+
+
+@lru_cache()
+def _qpos_tag(qubits: Union[cirq.LineQubit, Tuple[cirq.LineQubit]]):
+    """Given a qubit or qubits, return a "position tag" (used for drawing).
+
+    For multiple qubits, the tag is for the first qubit.
+    """
+    if isinstance(qubits, cirq.LineQubit):
+        return _qpos_tag((qubits,))
+    x = min(q.x for q in qubits)
+    return f'q{x}'
+
+
+@lru_cache()
+def _qpos_y(qubits: Union[cirq.LineQubit, Tuple[cirq.LineQubit]],
+            n_qubits: int):
+    """Given a qubit or qubits, return the position y value (used for drawing).
+
+    For multiple qubits, the position is the mean of the qubit indices.
+    This "flips" the coordinate so qubit 0 is at the maximal y position.
+    """
+    if isinstance(qubits, cirq.LineQubit):
+        return _qpos_y((qubits,), n_qubits)
+    x = np.mean([q.x for q in qubits]).item()
+    return n_qubits - x - 1
+
+
+def _add_to_positions(positions: Dict[Tuple[str, str], Tuple[float, float]],
+                      mi: int,
+                      qubits: Union[cirq.LineQubit, Tuple[cirq.LineQubit]], *,
+                      n_qubits: int, x_scale, y_scale, x_nudge, yb_offset):
+    """Helper function to update the `positions` dictionary.
+
+    Args:
+        positions: The dictionary to update. Quimb will consume this for drawing
+        mi: Moment index (used for x-positioning)
+        qubits: The qubits (used for y-positioning)
+        n_qubits: Total number of qubits, used for putting zero'th qubit on top
+        x_scale: Stretch coordinates in the x direction
+        y_scale: Stretch coordinates in the y direction
+        x_nudge: Kraus operators will have vertical lines connecting the
+            "forward" and "backward" circuits, so the x position of each
+            tensor is nudged (according to its y position) to help see all
+            the lines.
+        yb_offset: Offset the "backwards" circuit by this much.
+    """
+    qy = _qpos_y(qubits, n_qubits)
+    positions[(f'i{mi}f', _qpos_tag(qubits))] = \
+        (mi * x_scale + qy * x_nudge, y_scale * qy)
+    positions[(f'i{mi}b', _qpos_tag(qubits))] = \
+        (mi * x_scale, y_scale * qy + yb_offset)
+
+
+def circuit_to_density_matrix_tensors(circuit: cirq.Circuit,
+                                      qubits: Sequence[cirq.LineQubit]):
+    """Given a circuit with mixtures or channels, construct a tensor network
+    representation of the density matrix.
+
+    This assumes you start in the |0..0><0..| state. Indices are named
+    "nf{i}_q{x}" and "nb{i}_q{x}" where i is a time index and x is a
+    qubit index. nf- and nb- refer to the "forwards" and "backwards"
+    copies of the circuit. Kraus indices are named "k{j}" where j is an
+    independent "kraus" internal index which you probably never need to access.
+
+    Args:
+        circuit: The circuit containing operations that support the
+            cirq.unitary() or cirq.channel() protocols.
+        qubits: The qubits in the circuit.
```

Comment:
Can this be an optional argument, defaulting to `sorted(circuit.all_qubits())`?

Reply:
done here and below with test